### PR TITLE
[ROCm] Add hipDNN backend support for batch norm

### PR DESCRIPTION
**Pull request by zjgarvey**
### Motivation

This PR adds [hipDNN](https://github.com/ROCm/rocm-libraries/tree/develop/projects/hipdnn) batch norm support to PyTorch on ROCm, following the same dispatch op pattern used by cuDNN and MIOpen.

New ROCm kernel providers will be able to register plugins and ship faster kernels through hipDNN without touching PyTorch code. The graph-based API of hipDNN also supports for fused kernels. The eventual goal is to replace all MIOpen-specific PyTorch code with hipDNN, wherein MIOpen will be just one of multiple first-class kernel providers.

### Attribution

This work builds on Dmitry Nikolaev's ([@dnikolaev-amd](https://github.com/dnikolaev-amd)) hipDNN integration effort. Dmitry authored the foundational implementation on [ROCm/pytorch#3047](https://github.com/ROCm/pytorch/pull/3047) and opened an upstream draft at [#176917](https://github.com/pytorch/pytorch/pull/176917).

I'm opening a separate PR rather than pushing to Dmitry's branch so that I can iterate on reviewer feedback here while simultaneously driving an RFC discussion on [PyTorch Dev Discuss](https://dev-discuss.pytorch.org/t/rfc-adding-a-batch-norm-backend-revisiting-dispatch-stack-issues/3327) about improving batch norm's dispatch architecture more broadly. The RFC proposes completing the stalled `_batch_norm_with_update` migration ([#116092](https://github.com/pytorch/pytorch/pull/116092)), which would eventually eliminate the need for per-backend dispatch ops like the ones added here.

This PR incorporates Dmitry's original work plus fixes and additions from [ROCm/pytorch#3066](https://github.com/ROCm/pytorch/pull/3066) and [ROCm/pytorch#3068](https://github.com/ROCm/pytorch/pull/3068).

### Technical Details

**Integration approach:** Follows the established cuDNN/MIOpen per-backend dispatch op pattern: dedicated ops in `native_functions.yaml` with CUDA dispatch keys, autograd formulas in `derivatives.yaml`, decompositions for torch.compile/export, and vmap batching rules. We're aware this is the old-stack pattern and that `_batch_norm_with_update` (#116092) was designed to eliminate it. We chose to ship on the old stack because it's the only active code path today.

**New ops:**
- `hipdnn_batch_norm` (forward, training and inference)
- `hipdnn_batch_norm_backward` (backward)

**Backend selection:** Priority order is cuDNN → hipDNN → MIOpen → Native. hipDNN is **disabled by default** (`torch.backends.hipdnn.enabled = False`). Users opt in explicitly. When enabled, hipDNN takes priority over MIOpen because the user's intent is clear.

**Notable design choices:**

- **FC allowlist `9999-01-01`:** Both ops use permanent FC exemptions, matching the precedent set by `cudnn_batch_norm`, `miopen_batch_norm`, and other backend-specific ops in `check_forward_backward_compatibility.py`. These ops are needed as long as the old dispatch stack is active.

- **No graph caching (yet):** Each forward/backward call builds a new hipDNN frontend graph, allocates workspace, and executes. Graph caching is planned for a subsequent patch (e.g., see https://github.com/ROCm/pytorch/pull/3049). To focus the review process, this PR just adds the core hookup for a hipDNN integration: basic backend setup, build configuration, and support for one op.

- **New-stack forward-compatibility:** We added `BatchNormBackend::Hipdnn` cases to the dead `_batch_norm_with_update_cuda` / `_new_batch_norm_backward_cuda` code paths in `Normalization.cu`. This is untested (the new stack is unreachable), but ensures hipDNN is ready when the switch flips.

**Changes by area:**

| Area | Files | What |
|------|-------|------|
| CMake | `cmake/`, `aten/CMakeLists.txt` | Find and link hipDNN when available; `USE_HIPDNN` definition |
| Runtime hooks | `CUDAHooks`, `Context` | `compiledWithHipDNN()`, `userEnabledHipDNN()` |
| Dispatch ops | `native_functions.yaml`, `derivatives.yaml` | `hipdnn_batch_norm`, `hipdnn_batch_norm_backward` |
| Implementation | `aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp` | Forward/backward using hipDNN frontend graph API |
| Utilities | `aten/src/ATen/hipdnn/` | Handle management, type conversion, error handling, shared tensor attribute helpers |
| Backend selection | `Normalization.cpp`, `Normalization.h` | `BatchNormBackend::Hipdnn` case in `_select_batch_norm_backend` and `_batch_norm_impl_index` |
| New-stack compat | `Normalization.cu` | `Hipdnn` branches in dead `_batch_norm_with_update_cuda` code paths |
| Python backend | `torch/backends/hipdnn/` | `torch.backends.hipdnn.enabled` flag |
| Python enum | `Module.cpp` | `_BatchNormBackend.Hipdnn` |
| Decompositions | `decompositions.py` | Backward decomp to `native_batch_norm_backward` for torch.compile/export |
| vmap | `BatchRulesNorm.cpp`, `batch_norm_replacement.py` | Batching rules for hipdnn batch norm ops |
| FC compliance | `check_forward_backward_compatibility.py` | Permanent allowlist entries for new ops |

### Testing

Pytorch CI should be green, as hipDNN is disabled by default. The new hipDNN backend-plumbing ops should pass automated op checks.

The version of ROCm used in pytorch CI may not ship hipDNN at all, but we will be testing this code path nightly in [ROCm/TheRock](https://github.com/ROCm/TheRock).

### Related

- Forum Post: [Batch Norm RFC](https://dev-discuss.pytorch.org/t/rfc-adding-a-batch-norm-backend-revisiting-dispatch-stack-issues/3327)
- Dmitry's upstream PR: #176917
- ROCm integration: [ROCm/pytorch#3047](https://github.com/ROCm/pytorch/pull/3047)
- Fixes: [ROCm/pytorch#3066](https://github.com/ROCm/pytorch/pull/3066)
- Decomp fix: [ROCm/pytorch#3068](https://github.com/ROCm/pytorch/pull/3068)
- Batch norm consolidation (new stack): #116092
- Numerics divergence issue: #111384

### AI Tool Use

Claude Code (Claude Opus 4.6) was used extensively in my commits. I have changed "Co-Authored By:" trailers to "Assisted By:" to bypass automated CI checks. If there are established protocols here for providing transparency in AI use, please let me know.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @dnikolaev-amd

### CODE REVIEW

File: aten/src/ATen/hipdnn/Handle.cpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+namespace {
+
+void createHipdnnHandle(hipdnnHandle_t* handle) {
+  HIPDNN_CHECK(hipdnnCreate(handle));
+}
+
+void destroyHipdnnHandle(hipdnnHandle_t handle) {
+  // Intentionally not destroying handle to avoid shutdown ordering issues.
+  // See comments in the miopen equivalent (Handle.cpp).
+}
```
Comment (zjgarvey) [inline aten/src/ATen/hipdnn/Handle.cpp:19]: I personally do not have sufficient context for this decision. It seems to be established precedent, but I'd like to investigate if the original issue is resolved.

File: aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+#include <ATen/NativeFunctions.h>
+#else
+#include <ATen/ops/empty.h>
+#include <ATen/ops/empty_like.h>
+#include <ATen/ops/hipdnn_batch_norm_backward_native.h>
+#include <ATen/ops/hipdnn_batch_norm_native.h>
+#endif
+
+// TODO: Remove the condition on AT_ROCM_ENABLED entirely,
+// don't build this file as part of CPU build.
```
Comment (zjgarvey) [inline aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp:16]: Maybe add a PR link for context on the stalled migration plans.

File: aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+  while (static_cast<int64_t>(size.size()) < dim) {
+    size.emplace_back(1);
+  }
+  return t.view(size);
+}
+
+} // namespace
+
+std::tuple<Tensor, Tensor, Tensor> hipdnn_batch_norm(
+    const Tensor& input_t,
```
Comment (zjgarvey) [inline aten/src/ATen/native/hipdnn/BatchNorm_hipdnn.cpp:108]: I realize this is following precedent with parallel files, but I feel like `var_name_t` is not a good way to denote "var name tensor", as IDE's often think this is a type. Maybe we can make a case to deviate from this convention here.

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -4173,6 +4173,16 @@
     CUDA: miopen_batch_norm_backward
   autogen: miopen_batch_norm_backward.out
 
+- func: hipdnn_batch_norm(Tensor input, Tensor weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool training, float exponential_average_factor, float epsilon) -> (Tensor, Tensor, Tensor)
```
Comment (zjgarvey) [inline aten/src/ATen/native/native_functions.yaml:4176]: This should probably be added in alphabetical order, and not just copy/pasted next to miopen.

File: test/test_nn.py [hunk 5]
```diff
@@ -5252,7 +5250,7 @@ def test_batchnorm_buffer_update_when_stats_are_not_tracked(self):
         ],
         name_fn=lambda f, b, m, t: f"{f}_vs_{b}{'_mixed' if m else ''}_{dtype_name(t)}"
     )
-    def test_batchnorm(self, dims, mode, memory_format, ref_backend, mixed, dtype):
+    def test_batchnorm(self, backend, dims, mode, memory_format, ref_backend, mixed, dtype):
```
Comment (zjgarvey) [inline test/test_nn.py:5253]: This does automatically enable several batch norm tests for us, but seems like a temporary workaround to test hipDNN batch norm while the default is off. Maybe add a "#TODO" to remove the extra backend toggle when we make hipDNN default.

File: tools/autograd/derivatives.yaml [hunk 6]
```diff
@@ -2829,6 +2829,16 @@
 
 - name: mkldnn_rnn_layer_backward(Tensor input, Tensor weight1, Tensor weight2, Tensor weight3, Tensor weight4, Tensor hx_, Tensor cx_tmp, Tensor output, Tensor hy_, Tensor cy_, Tensor? grad_output, Tensor? grad_hy, Tensor? grad_cy, bool reverse, int mode, int hidden_size, int num_layers, bool has_biases, bool train, bool bidirectional, int[] batch_sizes, bool batch_first, Tensor workspace) -> (Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor)
 
+# hipdnn
+- name: hipdnn_batch_norm(Tensor input, Tensor weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool training, float exponential_average_factor, float epsilon) -> (Tensor, Tensor, Tensor)
```
Comment (zjgarvey) [inline tools/autograd/derivatives.yaml:2833]: Again, check for alphabetical order placement.

**Review by zjgarvey — COMMENTED** [review]
Small re-review comments before undraft.

File: tools/autograd/derivatives.yaml (hunk 6 above)
Comment (zjgarvey) [reply tools/autograd/derivatives.yaml:2833]: Doesn't appear to be an issue.

File: aten/src/ATen/native/Normalization.cpp [hunk 7]
```diff
@@ -520,6 +522,22 @@ BatchNormBackend _select_batch_norm_backend(
     return BatchNormBackend::Cudnn;
   }
 
+  // HipDNN — independent of MIOpen
+  if (at::globalContext().userEnabledHipdnn()
+      && detail::getCUDAHooks().compiledWithHipDNN()
+      && input.is_cuda()
+      && input.dim() >= 3
```
Comment (mousdahl-amd) [inline aten/src/ATen/native/Normalization.cpp:529]: Similar to the conv port, I can't guarantee these assertions will remain true forever.

Comment (zjgarvey) [reply aten/src/ATen/native/Normalization.cpp:529]: Yeah, this seems like a lot of conditions, honestly. I think Dmitry is looking into just building/caching the graph and querying hipDNN for available providers as a check, or we could iteratively update these whenever they can be relaxed.

Do you have any suggestions?

**Review by jeffdaily — APPROVED** [review]

**Comment by zjgarvey** [comment]
This is in draft state while we work on landing initial hipdnn design in #178515
